x/mongo/driver/crypt.go

```diff
@@ -382,3 +389,69 @@ func (c *crypt) decryptKey(kmsCtx *mongocrypt.KmsContext) error {
 		}
 	}
 }
+
+// needsKmsProvider returns true if provider was initially set to an empty document.
+// An empty document signals the driver to fetch credentials.
+func needsKmsProvider(kmsProviders *bsoncore.Document, provider string) bool {
+	gcpVal, err := kmsProviders.LookupErr(provider)
+	if err != nil {
+		// KMS provider is not configured.
+		return false
+	}
+	gcpDoc, ok := gcpVal.DocumentOK()
+	if !ok || len(gcpDoc) != 5 {
+		// KMS provider is not an empty document.
+		return false
+	}
+	return true
+}
+
+func (c *crypt) provideKmsProviders(cryptCtx *mongocrypt.Context) error {
+	kmsProviders := c.mongoCrypt.GetKmsProviders()
+	builder := bsoncore.NewDocumentBuilder()
+
+	if needsKmsProvider(kmsProviders, "gcp") {
+		// "gcp" KMS provider is an empty document.
+		// Attempt to fetch from GCP Instance Metadata server.
+		{
+			metadataHost := "metadata.google.internal"
+			if os.Getenv("GCE_METADATA_HOST") != "" {
+				metadataHost = os.Getenv("GCE_METADATA_HOST")
+			}
+			url := fmt.Sprintf("http://%s/computeMetadata/v1/instance/service-accounts/default/token", metadataHost)
```

Comment:
I think the safest way to do this is with the `net/url` package, something like this:

```go
u, err := url.Parse("computeMetadata/v1/instance/service-accounts/default/token")
if err != nil {
log.Fatal(err)
}
u.Scheme = "http"
u.Host = metadataHost
```

[Here](https://go.dev/play/p/BOSpoz_PNeI) is a playground example, and [here](https://pkg.go.dev/net/url#example-URL) is an example from the documentation.